pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -372,100 +367,80 @@ pub async fn disk_usage_eviction_task_iteration_impl<U: Usage>(
     };
     debug!(?usage_planned, "usage planned");
 
-    // phase2: evict victims batched by timeline
+    // phase2: evict layers
 
     let mut js = tokio::task::JoinSet::new();
+    let limit = 1000;
+
+    let mut evicted = candidates.into_iter().take(evicted_amount).fuse();
+    let mut consumed_all = false;
+
+    // After the evictions, `usage_assumed` is the post-eviction usage,
+    // according to internal accounting.
+    let mut usage_assumed = usage_pre;
+    let mut evictions_failed = LayerCount::default();
+
+    let join_all = async move {
+        loop {
+            let next = if js.len() >= limit || consumed_all {
+                js.join_next().await
+            } else if js.len() > limit / 10 {
+                // TODO: next tokio version exposes this as poll_join_next
```

Comment:
Aah, I actually already upgraded to that in #6061 :) This is a bit older code..

Reply:
I don't think the the poll variant actually helps now that I looked it more closely.

My aim here was to keep consuming the ready tasks while there are any but most of the time this would just join one or two until we are over the limit, and then pace the evicting..

I would not want to just fire up all of the evict_and_wait not that it would do any harm, but because that code would then be quite long blocking segment, and we really do not need to use that much memory anyways.. I'll try to make this more join all ready, so we will not need the yield_now either.

Reply:
I cannot actually find a better way to implement this; JoinSet poll_join_next could be consumed while it returns Poll::Ready, but it would mean another loop.

With the current implementation and the TODO comment removed, it will:
- consume the ready opportunistically after 10%
- pace next eviction after 1000 and flatline there; most likely new results will be ready instantly and we can continue
- after firing up all of the evictions we just sit waiting for the results

Reply:
It works, not perfectly pretty, but it works. I started thinking about cancelling by default after a timeout, but added only the bonus logging for this PR: [931511a](https://github.com/neondatabase/neon/pull/6060/commits/931511aecada44d18b1e5cc4ddd0eeab12c3901a)